python/dgl/frame.py

```diff
@@ -548,7 +559,14 @@ def record_stream(self, stream):
         """
         if F.get_preferred_backend() != "pytorch":
             raise DGLError("record_stream only supports the PyTorch backend.")
-        self.data.record_stream(stream)
+        # Only record streams for materialized tensors otherwise just record for the index
+        if self.index is None:
+            self.data.record_stream(stream)
```

Comment:
Want to point out that fix in this PR is not enough to avoid unnecessary H2D copies by just checking if `self.index` is materialized. In non-uva mode when `self.storage` is on cpu, the extra H2D copies can be launched from here. https://github.com/dmlc/dgl/blob/aa42aaeb9f499cce5af946f7e4aed15d7661cac8/python/dgl/frame.py#L247-L248

Reply:
Sorry, it seems these H2D copies in `data` method are necessary for fetching non-`lazy-feature`... right? @nv-dlasalle @yaox12 
Should look into it more closely :)

Reply:
@chang-l I think you were right, we dont want this function to cause any deferred actions to take place (`device` or `dtype`).

Reply:
I see... So, when `self.index` is `None`,  we should just have sth like below? Am I getting this correctly?   
```
if self.index is None:
    if F.context(self.storage) != F.cpu():
        self.storage.record_stream(stream)

Reply:
Correct.